### PR TITLE
multinomial newbackend

### PR DESCRIPTION
**Pull request by nouiz**

### CODE REVIEW

File: theano/sandbox/gpuarray/multinomial.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+from theano.sandbox import gpuarray
+from .basic_ops import as_gpuarray_variable, infer_context_name
+from .fp16_help import write_w
+from .opt import register_opt, op_lifter
+from .type import gpu_context_type, GpuArrayType
+
+
+class GPUAMultinomialFromUniform(gpuarray.basic_ops.GpuKernelBase, Op):
+    __props__ = ("odtype",)
+    params_type = gpu_context_type
```
Comment (abergeron) [inline theano/sandbox/gpuarray/multinomial.py]: This comes from GpuKernelBase.

File: theano/sandbox/gpuarray/multinomial.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        self.odtype = odtype
+
+    def get_params(self, node):
+        return node.outputs[0].type.context
+
+    def c_headers(self):
+        return ['<numpy_compat.h>', 'gpuarray_helper.h']
+
+    def c_header_dirs(self):
+        return [os.path.dirname(__file__), pygpu.get_include()]
```
Comment (abergeron) [inline theano/sandbox/gpuarray/multinomial.py]: The pygpu.get_include() is not mandatory is you use GpuArrays, you'll get it from the Type.

File: theano/sandbox/gpuarray/multinomial.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+import pygpu
+
+import theano
+import theano.sandbox.multinomial
+from theano import Apply, config
+from theano.gof import Op, local_optimizer
+from theano.tensor import NotScalarConstantError, get_scalar_constant_value
+from theano.sandbox import gpuarray
+from .basic_ops import as_gpuarray_variable, infer_context_name
+from .fp16_help import write_w
```
Comment (abergeron) [inline theano/sandbox/gpuarray/multinomial.py]: Why would you need this if you restrict support to float32 in the make_node?

File: theano/sandbox/gpuarray/multinomial.c [hunk 4]
```diff
@@ -0,0 +1,135 @@
+#section support_code_apply
+
```
Comment (abergeron) [inline theano/sandbox/gpuarray/multinomial.c]: This whole file is unused.

File: theano/sandbox/gpuarray/multinomial.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+            raise NotImplementedError('unis ndim should be 1', unis.ndim)
+        if self.odtype == 'auto':
+            odtype = pvals.dtype
+        else:
+            odtype = self.odtype
+        assert odtype == 'float32', odtype
+        if odtype != pvals.dtype:
+            raise NotImplementedError(
+                'GpuMultinomialFromUniform works only if '
+                'self.odtype == pvals.dtype', odtype, pvals.dtype)
```
Comment (abergeron) [inline theano/sandbox/gpuarray/multinomial.py]: I'm just wondering why do you allow setting odtype, if in the end anything that is not the same as the dtype of pvals will give you an error?

File: theano/sandbox/gpuarray/multinomial.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+        br = (pvals.broadcastable[1], pvals.broadcastable[0])
+        out = GpuArrayType(broadcastable=br, dtype=odtype)()
```
Comment (abergeron) [inline theano/sandbox/gpuarray/multinomial.py]: You need to pass the context_name to the GpuArrayType constructor.

File: theano/sandbox/gpuarray/multinomial.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+        }
+    }
+}
+
+
+//KERNEL void k(GLOBAL_MEM %(ctype)s *a, ga_size n, ga_size m) {
+//    ga_size nb = n < m ? n : m;
+//    for (ga_size i = LID_0; i < nb; i += LDIM_0) {
+//        a[i*m + i] = %(write_a)s(1);
+//    }
```
Comment (abergeron) [inline theano/sandbox/gpuarray/multinomial.py]: Maybe you shouldn't keep a commented-out kernel here?

File: theano/sandbox/gpuarray/multinomial.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+//}""" % dict(ctype=pygpu.gpuarray.dtype_to_ctype(dtype),
+              name=name, write_a=write_w(dtype))
```
Comment (abergeron) [inline theano/sandbox/gpuarray/multinomial.py]: `write_a` and `ctype` are only used for the commented kernel.  `name` is not used anywhere.

File: theano/sandbox/gpuarray/multinomial.py [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+                ],
+            flags=gpuarray.basic_ops.Kernel.get_flags(node.outputs[0].dtype),
+            objvar='k_multi_warp_multinomial_' + name)]
+
+    def c_code(self, node, name, inp, outputs, sub):
+        pvals, unis = inp
+        out, = outputs
+        fail = sub['fail']
+        ctx = sub['params']
+        #typecode = pygpu.gpuarray.dtype_to_typecode(self.dtype)
```
Comment (abergeron) [inline theano/sandbox/gpuarray/multinomial.py]: If you don't need it, then don't keep it as a comment.

File: theano/sandbox/gpuarray/multinomial.py [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+                           GA_C_ORDER, %(ctx)s) != 0){
+      %(fail)s
+    }
+    %(out)s = out;
+    GpuArray_memset(&(out->ga), 0);
+    { // NESTED SCOPE
+        int nb_multi = PyGpuArray_DIMS(pvals)[0];
+        int nb_outcomes = PyGpuArray_DIMS(pvals)[1];
+        //TODO : change this for a beautiful constant
+        int max_nb_blocks = 2<<15 - 1;
```
Comment (abergeron) [inline theano/sandbox/gpuarray/multinomial.py]: You can actually query this value from the kernel to have the real maximum number of blocks.

File: theano/sandbox/gpuarray/tests/test_multinomial.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    mode = get_default_mode()
+    mode = copy.copy(mode)
+    if gpu:
+        mode = mode.including('gpuarray', 'gpu_local_optimizations',
+                              'local_cut_gpu_host_gpu')
+    if isinstance(mode.linker, theano.gof.PerformLinker):
+        mode.linker = predefined_linkers['c|py']
+    if hasattr(mode.linker, 'c_thunks'):
+        mode.linker.c_thunks = True
+    return mode
```
Comment (abergeron) [inline theano/sandbox/gpuarray/tests/test_multinomial.py]: What is this?  What does it performs black magic on the mode?

File: theano/sandbox/gpuarray/multinomial.py (hunk 3 above)
Comment (nouiz) [inline theano/sandbox/gpuarray/multinomial.py]: Part of the multi-dtype, float16 support it done, but not all. I let the finishing of this to another PR, so we can deprecate the old back-end earlier.

File: theano/sandbox/gpuarray/multinomial.py (hunk 9 above)
Comment (nouiz) [inline theano/sandbox/gpuarray/multinomial.py]: Will be needed later when we support multiple-dtype.

File: theano/sandbox/gpuarray/multinomial.py (hunk 10 above)
Comment (nouiz) [inline theano/sandbox/gpuarray/multinomial.py]: Just porting code, I won't do that now.

File: theano/sandbox/gpuarray/multinomial.py (hunk 1 above)
Comment (nouiz) [inline theano/sandbox/gpuarray/multinomial.py]: removed

File: theano/sandbox/gpuarray/multinomial.py (hunk 2 above)
Comment (nouiz) [inline theano/sandbox/gpuarray/multinomial.py]: removed

File: theano/sandbox/gpuarray/multinomial.c (hunk 4 above)
Comment (nouiz) [inline theano/sandbox/gpuarray/multinomial.c]: removed

**Comment by nouiz** [comment]
I fixed your comment and fixed the way the old test was handling the mode.

File: theano/sandbox/gpuarray/multinomial.py (hunk 3 above)
Comment (abergeron) [inline theano/sandbox/gpuarray/multinomial.py]: But you don't actually use it anywhere.  I'm not against making float16 support in a follow-up pass, but don't import unused stuff.

File: theano/sandbox/gpuarray/multinomial.py (hunk 9 above)
Comment (abergeron) [inline theano/sandbox/gpuarray/multinomial.py]: Then we can add it later.

File: theano/sandbox/gpuarray/multinomial.py (hunk 10 above)
Comment (abergeron) [inline theano/sandbox/gpuarray/multinomial.py]: Ok

File: theano/sandbox/gpuarray/tests/test_multinomial.py [hunk 12]
```diff
@@ -14,14 +12,12 @@
...
+        mode = theano.compile.get_mode('FAST_RUN')
     if gpu:
         mode = mode.including('gpuarray', 'gpu_local_optimizations',
-                              'local_cut_gpu_host_gpu')
-    if isinstance(mode.linker, theano.gof.PerformLinker):
-        mode.linker = predefined_linkers['c|py']
-    if hasattr(mode.linker, 'c_thunks'):
-        mode.linker.c_thunks = True
+                              'local_cut_gpu_host_gpu',
+                              'local_gpu_multinomial')
```
Comment (abergeron) [inline theano/sandbox/gpuarray/tests/test_multinomial.py]: Why not use the usual mode_with_gpu/mode_without_gpu?

**Comment by abergeron** [comment]
There is a bunch of flake8 breakage:
```
/home/travis/build/Theano/Theano/theano/sandbox/gpuarray/multinomial.py:1:1: FI10 __future__ import "division" missing

/home/travis/build/Theano/Theano/theano/sandbox/gpuarray/multinomial.py:1:1: FI11 __future__ import "absolute_import" missing

/home/travis/build/Theano/Theano/theano/sandbox/gpuarray/multinomial.py:1:1: FI13 __future__ import "print_function" missing

/home/travis/build/Theano/Theano/theano/sandbox/gpuarray/multinomial.py:9:1: F401 'local_optimizer' imported but unused

/home/travis/build/Theano/Theano/theano/sandbox/gpuarray/multinomial.py:13:1: F401 'write_w' imported but unused

/home/travis/build/Theano/Theano/theano/sandbox/gpuarray/multinomial.py:15:1: F401 'gpu_context_type' imported but unused

/home/travis/build/Theano/Theano/theano/sandbox/gpuarray/multinomial.py:63:9: F841 local variable 'dtype' is assigned to but never used

/home/travis/build/Theano/Theano/theano/sandbox/gpuarray/multinomial.py:117:17: E124 closing bracket does not match visual indentation

/home/travis/build/Theano/Theano/theano/sandbox/gpuarray/multinomial.py:126:9: E265 block comment should start with '# '

/home/travis/build/Theano/Theano/theano/sandbox/gpuarray/tests/test_multinomial.py:8:1: F401 'predefined_linkers' imported but unused

/home/travis/build/Theano/Theano/theano/sandbox/gpuarray/tests/test_multinomial.py:10:1: F401 'mode_with_gpu' imported but unused

/home/travis/build/Theano/Theano/theano/sandbox/gpuarray/tests/test_multinomial.py:10:1: F401 'mode_without_gpu' imported but unused

/home/travis/build/Theano/Theano/theano/sandbox/gpuarray/tests/test_multinomial.py:10:1: F401 'test_ctx_name' imported but unused

/home/travis/build/Theano/Theano/theano/sandbox/tests/test_multinomial.py:12:1: F401 'predefined_linkers' imported but unused
```

In addition to that, you have an unprotected import of pygpu in multinomial.py which makes the import fail on travis.

**Comment by nouiz** [comment]
I fixed the comments

**Comment by abergeron** [comment]
Ok when travis passes

**Comment by nouiz** [comment]
travis was always failing with a timeout at the same test in FAST_COMPILE:

theano/tensor/nnet/tests/test_bn.py:test_bn_feature_maps

I have rebased on master just in case it help. I'm checking locally why this test is so slow.

**Pull request merged**